osquery/tables/networking/windows/windows_firewall_rules.cpp

```diff
@@ -0,0 +1,411 @@
+    IEnumVARIANT* enumvar, bool& found) {
+  found = false;
+
+  WindowsFirewallRule winFwRule;
+  Expected<WindowsFirewallRule, WindowsFirewallError> r(winFwRule);
+
+  ULONG fetched = 0;
+
+  VARIANT value;
+  VariantInit(&value);
+
+  INetFwRule* rule = NULL;
+
+  HRESULT hr = enumvar->Next(1, &value, &fetched);
+  if (FAILED(hr)) {
+    r = createWindowsFirewallRuleError(
+        WindowsFirewallError::PolicyRulesEnumNextError, hr);
+  } else {
+    if (hr != S_FALSE) {
+      found = true;
+      hr = VariantChangeType(&value, &value, 0, VT_DISPATCH);
+      if (FAILED(hr)) {
+        // Log only
+        auto err = createWindowsFirewallRuleError(
+            WindowsFirewallError::PolicyRulesEnumIDispatchError, hr);
+        TLOG << err.getError().getMessage();
+      } else {
+        hr = (V_DISPATCH(&value))->QueryInterface(IID_PPV_ARGS(&rule));
+        if (FAILED(hr)) {
+          // Log only
+          auto err = createWindowsFirewallRuleError(
+              WindowsFirewallError::PolicyRuleInterfaceError, hr);
+          TLOG << err.getError().getMessage();
+        } else {
+          r = populateFirewallRule(rule);
+        }
+      }
+    }
+  }
+
+  // Cleanup
+  if (rule != NULL) {
+    rule->Release();
+  }
+
+  VariantClear(&value);
+
+  return r;
+}
+
+Expected<WindowsFirewallRules, WindowsFirewallError> getFirewallRules(
+    QueryContext& context) {
+  WindowsFirewallRules results;
+  Expected<WindowsFirewallRules, WindowsFirewallError> r(results);
+
+  INetFwPolicy2* netFwPolicy = NULL;
+  INetFwRules* rules = NULL;
+  IUnknown* enumerator = NULL;
+  IEnumVARIANT* enumvar = NULL;
+
+  HRESULT hr = CoCreateInstance(__uuidof(NetFwPolicy2),
+                                NULL,
+                                CLSCTX_INPROC_SERVER,
+                                IID_PPV_ARGS(&netFwPolicy));
+  if (FAILED(hr)) {
+    r = createWindowsFirewallRulesError(WindowsFirewallError::NetFwPolicyError,
+                                        hr);
+    goto Cleanup;
+  }
+
+  hr = netFwPolicy->get_Rules(&rules);
+  if (FAILED(hr)) {
+    r = createWindowsFirewallRulesError(WindowsFirewallError::PolicyRulesError,
+                                        hr);
+    goto Cleanup;
+  }
+
+  hr = rules->get__NewEnum(&enumerator);
+  if (FAILED(hr)) {
+    r = createWindowsFirewallRulesError(
+        WindowsFirewallError::PolicyRulesEnumError, hr);
+    goto Cleanup;
+  }
+
+  hr = enumerator->QueryInterface(IID_PPV_ARGS(&enumvar));
+  if (FAILED(hr)) {
+    r = createWindowsFirewallRulesError(
+        WindowsFirewallError::PolicyRulesEnumInterfaceError, hr);
+    goto Cleanup;
+  }
+
+  bool found = true;
+  while (found) {
+    auto res = getNext(enumvar, found);
+    if (res.isError()) {
+      r = res.takeError();
+      break;
+    }
+    if (found) {
+      results.push_back(std::move(res.get()));
+    }
+  }
+
+  if (!r.isError()) {
+    r = results;
+  }
+
+Cleanup:
+
+  if (enumvar != NULL) {
```

Comment:
Manual cleanup should be avoided (https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#e6-use-raii-to-prevent-leaks).

I think these types can be easily wrapped around `std::unique_ptr` with a custom deleter. An example can be found here: `https://github.com/osquery/osquery/blob/master/osquery/tables/system/windows/intel_me.cpp#L260`